File: packages/react-native-yubikey/ios/HybridYubiKeyPiv.swift
import CoreNFC
import Foundation
import YubiKit

/**
 * YubiKeyPiv over Yubico's YubiKit, NFC transport (`YKFNFCConnection` +
 * `YKFPIVSession`).
 *
 * WHY NFC and not USB-C on iOS: the USB-C smart-card path
 * (`YKFSmartCardConnection`) needs `com.apple.security.smartcard`, a macOS App
 * Sandbox entitlement that iOS App Store validation rejects. NFC ISO7816 is the
 * only App-Store-valid way for a third-party app to reach a YubiKey's PIV
 * applet on iOS, via the `com.apple.developer.nfc.readersession.formats` (TAG)
 * capability + the iso7816 select-identifiers in Info.plist. It requires an
 * NFC-capable key (e.g. YubiKey 5C NFC). Android keeps USB-C CCID.
 *
 * NFC lifecycle: unlike a persistent USB reader, an NFC session is a modal tap
 * — `startNFCConnection()` shows the system scan sheet, the user holds the key
 * to the top of the phone, `didConnectNFC` fires, and the connection stays open
 * (so a whole ceremony — verify PIN then touch-gated ECDH — runs in ONE tap)
 * until `stopNFCConnection()`. So the JS layer calls start() only when a
 * ceremony begins (never at launch), and stop() when it arms or fails.
 *
 * Every rejection carries a `VAULT_ERR:<code>:<detail>` message (see `mapError`)
 * so the JS vault layer can branch on a stable machine code. Completion handlers
 * are all guarded — a dropped tap or a YubiKit error becomes a rejection.
 */
final class HybridYubiKeyPiv: HybridYubiKeyPivSpec {
  /// (eventType, serial, transport) -> Void. Nitro dispatches the call to JS.
  private var listener: ((String, String, String) -> Void)?
  /// The connection for the key currently on a reader. Set on didConnect*,
  /// cleared on didDisconnect*. Every operation runs against this.
  fileprivate var activeConnection: (any YKFConnectionProtocol)?
  /// YKFManagerDelegate requires an NSObject conformer, which this Nitro
  /// HybridObject is not — so the delegate lives on a separate NSObject that
  /// forwards connect/disconnect back here. Held strong; YubiKitManager keeps
  /// only a weak reference.
  private lazy var connDelegate: ConnectionDelegate = {
    let d = ConnectionDelegate()
    d.owner = self
    return d
  }()

  // MARK: - Capability

  /// NFC ISO7816 needs iOS 13+ and NFC hardware. There is no App-Store-valid
  /// USB-C smart-card path on iOS, so when NFC is unavailable the JS layer
  /// treats the device as reader-less and stays on its software-key path.
  func isSupported() throws -> Bool {
    if #available(iOS 13, *) { return NFCReaderSession.readingAvailable } else { return false }
  }

  // MARK: - Discovery (an NFC tap)

  /// Begin an NFC session — the JS layer calls this only when a ceremony needs a
  /// key, NEVER at launch (it presents the system scan sheet). The session stays
  /// open across the whole ceremony until stopDiscovery().
  func startDiscovery() throws {
    YubiKitManager.shared.delegate = connDelegate
    YubiKitExternalLocalization.nfcScanAlertMessage =
      "Hold your YubiKey to the top of your phone to unlock the vault."
    if #available(iOS 13.0, *) {
      YubiKitManager.shared.startNFCConnection()
    }
  }

  /// End the NFC session (dismisses the scan sheet). Called on arm / terminal
  /// error / cancel by the JS ceremony.
  func stopDiscovery() throws {
    if #available(iOS 13.0, *) {
      YubiKitManager.shared.stopNFCConnection()
    }
    YubiKitManager.shared.delegate = nil
    activeConnection = nil
  }

  func setKeyListener(listener: @escaping (String, String, String) -> Void) throws {
    self.listener = listener
  }

  func clearKeyListener() throws {
    self.listener = nil
  }

  fileprivate func emit(_ eventType: String, _ serial: String, _ transport: String) {
    listener?(eventType, serial, transport)
  }

  /// Called by the delegate on a connect: hold the connection and emit.
  fileprivate func handleConnect(_ connection: any YKFConnectionProtocol, _ transport: String) {
    activeConnection = connection
    readSerialAndEmit(connection, transport)
  }

  /// Called by the delegate on a disconnect: drop the connection if it is the
  /// one we hold, and emit a removal so the JS layer relocks.
  fileprivate func handleDisconnect(_ connection: AnyObject, _ transport: String) {
    if (activeConnection as AnyObject?) === connection { activeConnection = nil }
    emit("removed", "", transport)
  }

  /// Opens a throwaway session just to read the serial for a connect event.
  private func readSerialAndEmit(_ connection: any YKFConnectionProtocol, _ transport: String) {
    connection.pivSession { [weak self] session, _ in
      guard let self else { return }
      guard let session else { self.emit("connected", "", transport); return }
      session.getSerialNumber { serial, error in
        let serialStr = error == nil ? String(serial) : ""
        self.emit("connected", serialStr, transport)
      }
    }
  }

  // MARK: - Operations

  func getKeyInfo() throws -> Promise<String> {
    let promise = Promise<String>()
    withSession(promise) { session in
      let version = session.version
      session.getPinAttempts { attempts, _ in
        session.getSerialNumber { serial, error in
          if let error { return promise.reject(withError: Self.mapError(error)) }
          let json = "{\"serial\":\"\(serial)\",\"firmwareVersion\":\"\(version.major).\(version.minor).\(version.micro)\",\"pinRetries\":\(attempts)}"
          promise.resolve(withResult: json)
        }
      }
    }
    return promise
  }

  func verifyPin(pin: String) throws -> Promise<String> {
    let promise = Promise<String>()
    withSession(promise) { session in
      session.verifyPin(pin) { retriesLeft, error in
        // A wrong PIN is a normal, resolvable result for this probe (the spec
        // returns {ok, retriesLeft}); only transport faults reject.
        if error != nil {
          promise.resolve(withResult: "{\"ok\":false,\"retriesLeft\":\(retriesLeft)}")
        } else {
          promise.resolve(withResult: "{\"ok\":true,\"retriesLeft\":null}")
        }
      }
    }
    return promise
  }

  func changePin(oldPin: String, newPin: String) throws -> Promise<String> {
    let promise = Promise<String>()
    withSession(promise) { session in
      // Per the module spec, changePin is grouped with generateKey under the
      // management-key gate. (PIV's CHANGE REFERENCE DATA itself only needs the
      // old PIN; the management-key auth is here because the spec asks for it,
      // and it is what surfaces mgmt-key-custom on a personalised key.)
      self.authenticateManagementKey(session, promise) {
        session.setPin(newPin, oldPin: oldPin) { error in
          if let error { return promise.reject(withError: Self.mapError(error)) }
          promise.resolve(withResult: "{\"ok\":true}")
        }
      }
    }
    return promise
  }

  func generateVaultKey(slot: Double, touchPolicy: String, pinPolicy: String) throws -> Promise<String> {
    let promise = Promise<String>()
    guard let pivSlot = YKFPIVSlot(rawValue: UInt(slot)) else {
      promise.reject(withError: Self.vaultError("no-key", "bad slot"))
      return promise
    }
    withSession(promise) { session in
      self.authenticateManagementKey(session, promise) {
        session.generateKey(
          in: pivSlot,
          type: .ECCP256,
          pinPolicy: Self.toPinPolicy(pinPolicy),
          touchPolicy: Self.toTouchPolicy(touchPolicy)
        ) { publicKey, error in
          if let error { return promise.reject(withError: Self.mapError(error)) }
          guard let publicKey, let hex = Self.secKeyToSec1Hex(publicKey) else {
            return promise.reject(withError: Self.vaultError("wrong-key", "could not export public key"))
          }
          promise.resolve(withResult: "{\"publicKey\":\"\(hex)\"}")
        }
      }
    }
    return promise
  }

  func readVaultPublicKey(slot: Double) throws -> Promise<String> {
    let promise = Promise<String>()
    guard let pivSlot = YKFPIVSlot(rawValue: UInt(slot)) else {
      promise.reject(withError: Self.vaultError("no-key", "bad slot"))
      return promise
    }
    withSession(promise) { session in
      // YubiKit 4.4 exposes no slot key-metadata read (getSlotMetadata was added
      // later / is Android-only here). We use the slot CERTIFICATE as the
      // occupancy signal: PIV tooling that owns a retired slot (age-plugin-
      // yubikey, ykman) writes an X.509 cert alongside the key, so a present
      // cert means "occupied — don't overwrite". Limitation: a bare keypair
      // with no cert reads as empty on iOS; the enrollment flow only generates
      // over confirmed-empty slots, and this still protects the common case.
      session.getCertificateIn(pivSlot) { certificate, error in
        guard error == nil, let certificate,
              let pub = SecCertificateCopyKey(certificate),
              let hex = Self.secKeyToSec1Hex(pub) else {
          return promise.resolve(withResult: "{\"publicKey\":null}")
        }
        promise.resolve(withResult: "{\"publicKey\":\"\(hex)\"}")
      }
    }
    return promise
  }

  func ecdh(slot: Double, pin: String, peerPublicKey: String) throws -> Promise<String> {
    let promise = Promise<String>()
    guard let pivSlot = YKFPIVSlot(rawValue: UInt(slot)) else {
      promise.reject(withError: Self.vaultError("no-key", "bad slot"))
      return promise
    }
    guard let peerKey = Self.sec1HexToSecKey(peerPublicKey) else {
      promise.reject(withError: Self.vaultError("wrong-key", "bad peer public key"))
      return promise
    }
    withSession(promise) { session in
      // pin-policy ONCE gate: verify before the touch-gated agreement.
      session.verifyPin(pin) { _, error in
        if let error { return promise.reject(withError: Self.mapError(error)) }
        // TOUCH-gated when the key was generated with TouchPolicy.ALWAYS: this
        // waits for the user to tap the key, and surfaces as touch-timeout if
        // they never do (see mapError).
        session.calculateSecretKey(in: pivSlot, peerPublicKey: peerKey) { secret, error in
          if let error { return promise.reject(withError: Self.mapError(error)) }
          guard let secret else {
            return promise.reject(withError: Self.vaultError("touch-timeout", "no shared secret returned"))
          }
          promise.resolve(withResult: "{\"secret\":\"\(secret.hexString)\"}")
        }
      }
    }
    return promise
  }

  // MARK: - Helpers

  /// Opens a `YKFPIVSession` on the held connection and hands it to `work`;
  /// rejects with no-key when nothing is on a reader, or maps a session-open
  /// error. `work` owns resolving/rejecting `promise` from there.
  private func withSession(_ promise: Promise<String>, _ work: @escaping (YKFPIVSession) -> Void) {
    guard let connection = activeConnection else {
      promise.reject(withError: Self.vaultError("no-key", "no YubiKey present"))
      return
    }
    connection.pivSession { session, error in
      if let error { return promise.reject(withError: Self.mapError(error)) }
      guard let session else {
        return promise.reject(withError: Self.vaultError("no-key", "could not open PIV session"))
      }
      work(session)
    }
  }

  /// Authenticate with the firmware-default management key so generateKey (and,
  /// per spec, changePin) can proceed, then run `next`. Pre-5.7 keys default to
  /// TDES, fw >= 5.7 to AES-192; both ship the same 24-byte default value. A
  /// failure means a custom management key we cannot supply → mgmt-key-custom.
  private func authenticateManagementKey(
    _ session: YKFPIVSession,
    _ promise: Promise<String>,
    _ next: @escaping () -> Void
  ) {
    let version = session.version
    let fw57 = version.major > 5 || (version.major == 5 && version.minor >= 7)
    // 4.4: these are class factory methods on YKFPIVManagementKeyType, so call
    // them (they are not enum cases).
    let type: YKFPIVManagementKeyType = fw57 ? .aes192() : .tripleDES()
    session.authenticate(withManagementKey: Self.defaultManagementKey, type: type) { error in
      if error != nil {
        return promise.reject(withError: Self.vaultError("mgmt-key-custom", "default management key rejected"))
      }
      next()
    }
  }

  private static func toTouchPolicy(_ p: String) -> YKFPIVTouchPolicy {
    switch p.lowercased() {
    case "always": return .always
    case "cached": return .cached
    case "never": return .never
    default: return .default
    }
  }

  private static func toPinPolicy(_ p: String) -> YKFPIVPinPolicy {
    switch p.lowercased() {
    case "once": return .once
    case "always": return .always
    case "never": return .never
    default: return .default
    }
  }

  private static func vaultError(_ code: String, _ detail: String) -> NSError {
    NSError(domain: "YubiKeyPiv", code: 1,
            userInfo: [NSLocalizedDescriptionKey: "VAULT_ERR:\(code):\(detail)"])
  }

  /// Best-effort translation of a YubiKit error to a VAULT_ERR code.
  ///
  /// NOTE: YubiKit surfaces smart-card faults as NSErrors whose `code` is often
  /// the raw APDU status word; the exact domains/codes are version-sensitive
  /// and should be re-checked on-device. Anything unrecognised falls through to
  /// wrong-key so a failure is never silently swallowed.
  private static func mapError(_ error: Error) -> NSError {
    let ns = error as NSError
    // Already one of ours (e.g. nested through withSession) — pass through.
    if ns.domain == "YubiKeyPiv" { return ns }
    switch ns.code {
    case 0x6983: return vaultError("pin-locked", "authentication method blocked")
    case 0x63C0...0x63CF: // 0x63Cx = PIN verify failed, x = retries left
      return vaultError("pin-invalid", "retries=\(ns.code & 0x0F)")
    case 0x6A88, 0x6A80: return vaultError("no-key", "reference data not found")
    case 0x6982, 0x6985: return vaultError("touch-timeout", "conditions of use not satisfied")
    default:
      let desc = ns.localizedDescription.lowercased()
      if desc.contains("touch") || desc.contains("timeout") {
        return vaultError("touch-timeout", ns.localizedDescription)
      }
      if desc.contains("no connection") || desc.contains("disconnect") || desc.contains("removed") {
        return vaultError("key-removed-mid-op", ns.localizedDescription)
      }
      return vaultError("wrong-key", ns.localizedDescription)
    }
  }

  /// EC public `SecKey` -> SEC1 uncompressed hex (0x04 || X || Y). Security's
  /// external representation for an EC public key IS ANSI X9.63 uncompressed,
  /// so this is a straight export + hex encode.
  private static func secKeyToSec1Hex(_ key: SecKey) -> String? {
    var error: Unmanaged<CFError>?
    guard let data = SecKeyCopyExternalRepresentation(key, &error) as Data? else { return nil }
    return data.hexString
  }

  /// SEC1 uncompressed hex (0x04 || X || Y) -> EC public `SecKey`.
  private static func sec1HexToSecKey(_ hex: String) -> SecKey? {
    guard let data = Data(hexString: hex) else { return nil }
    let attrs: [String: Any] = [
      kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
      kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
      kSecAttrKeySizeInBits as String: 256
    ]
    var error: Unmanaged<CFError>?
    return SecKeyCreateWithData(data as CFData, attrs as CFDictionary, &error)
  }

  /// Firmware-default PIV management key (0x0102…08 ×3, 24 bytes).
  private static let defaultManagementKey = Data([
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
  ])
}

// MARK: - Discovery delegate

/// YKFManagerDelegate is declared `<NSObject>`, so its conformer must be an
/// NSObject — which the Nitro HybridObject is not. This lightweight NSObject
/// holds a weak back-reference and forwards every connect/disconnect to the
/// owner. The SmartCard callbacks carry no availability annotation in YubiKit
/// 4.4's protocol, so none is needed here (startSmartCardConnection, which is
/// iOS 16+, is already guarded at the call site).
private final class ConnectionDelegate: NSObject, YKFManagerDelegate {
  weak var owner: HybridYubiKeyPiv?

  func didConnectNFC(_ connection: YKFNFCConnection) {
    owner?.handleConnect(connection, "nfc")
  }
  func didDisconnectNFC(_ connection: YKFNFCConnection, error: Error?) {
    owner?.handleDisconnect(connection, "nfc")
  }
  func didConnectAccessory(_ connection: YKFAccessoryConnection) {
    owner?.handleConnect(connection, "usb")
  }
  func didDisconnectAccessory(_ connection: YKFAccessoryConnection, error: Error?) {
    owner?.handleDisconnect(connection, "usb")
  }
  func didConnectSmartCard(_ connection: YKFSmartCardConnection) {
    owner?.handleConnect(connection, "usb")
  }
  func didDisconnectSmartCard(_ connection: YKFSmartCardConnection, error: Error?) {
    owner?.handleDisconnect(connection, "usb")
  }
}

// MARK: - hex

private extension Data {
  var hexString: String { map { String(format: "%02x", $0) }.joined() }

  init?(hexString: String) {
    var hex = hexString
    if hex.hasPrefix("0x") || hex.hasPrefix("0X") { hex = String(hex.dropFirst(2)) }
    guard hex.count % 2 == 0 else { return nil }
    var out = Data(capacity: hex.count / 2)
    var idx = hex.startIndex
    while idx < hex.endIndex {
      let next = hex.index(idx, offsetBy: 2)
      guard let byte = UInt8(hex[idx..<next], radix: 16) else { return nil }
      out.append(byte)
      idx = next
    }
    self = out
  }
}

File: packages/react-native-yubikey/android/src/main/java/com/margelo/nitro/yubikeypiv/HybridYubiKeyPiv.kt
package com.margelo.nitro.yubikeypiv

import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.yubico.yubikit.android.YubiKitManager
import com.yubico.yubikit.android.transport.nfc.NfcConfiguration
import com.yubico.yubikit.android.transport.nfc.NfcNotAvailable
import com.yubico.yubikit.android.transport.nfc.NfcYubiKeyDevice
import com.yubico.yubikit.android.transport.usb.UsbConfiguration
import com.yubico.yubikit.android.transport.usb.UsbYubiKeyDevice
import com.yubico.yubikit.core.YubiKeyDevice
import com.yubico.yubikit.core.keys.EllipticCurveValues
import com.yubico.yubikit.core.keys.PublicKeyValues
import com.yubico.yubikit.core.smartcard.ApduException
import com.yubico.yubikit.core.smartcard.SmartCardConnection
// 3.2.0: InvalidPinException lives in core.application, not piv.
import com.yubico.yubikit.core.application.InvalidPinException
import com.yubico.yubikit.piv.KeyType
import com.yubico.yubikit.piv.PinPolicy
import com.yubico.yubikit.piv.PivSession
import com.yubico.yubikit.piv.Slot
import com.yubico.yubikit.piv.TouchPolicy
import java.io.IOException

/**
 * YubiKeyPiv over YubiKit-Android's PIV application (CCID).
 *
 * Mirrors the iOS/YubiKit backend's contract exactly (the JS wrapper is
 * shared): discovery streams key-connected / key-removed events to the JS
 * listener, and each Promise-returning method opens a fresh
 * SmartCardConnection → PivSession against the currently-held device, does its
 * one operation, and lets the connection close.
 *
 * Every rejection carries a `VAULT_ERR:<code>:<detail>` message (see
 * `mapError`) so the JS vault layer can branch on a stable machine code rather
 * than parse YubiKit's own English.
 *
 * Discovery bookkeeping (`listener`, `currentDevice`) is confined to the
 * main-thread Handler — YubiKitManager delivers USB/NFC discovery on the main
 * thread and we post everything we initiate onto the same thread, making the
 * discovery state machine single-threaded by construction (the same discipline
 * the Swift side gets from serial completion queues). The per-operation
 * `requestConnection` callbacks run on YubiKit's own executor; they only touch
 * the Promise they were handed, never the shared discovery state.
 */
class HybridYubiKeyPiv : HybridYubiKeyPivSpec() {
  private val main = Handler(Looper.getMainLooper())

  private val manager: YubiKitManager? by lazy {
    val ctx = NitroModules.applicationContext ?: return@lazy null
    YubiKitManager(ctx.applicationContext)
  }

  /** JS listener: (eventType, serial, transport). Confined to `main`. */
  private var listener: ((String, String, String) -> Unit)? = null
  /** The key currently on a reader (USB plugged, or NFC held). Confined to `main`. */
  private var currentDevice: YubiKeyDevice? = null
  private var discovering = false

  // ── discovery ──

  override fun isSupported(): Boolean {
    val ctx = NitroModules.applicationContext ?: return false
    val pm: PackageManager = ctx.packageManager
    val usb = pm.hasSystemFeature(PackageManager.FEATURE_USB_HOST)
    val nfc = pm.hasSystemFeature(PackageManager.FEATURE_NFC)
    return usb || nfc
  }

  override fun startDiscovery() {
    main.post {
      if (discovering) return@post
      val m = manager ?: return@post
      discovering = true

      // USB: the SDK owns the runtime permission dialog. Each plug-in delivers
      // a UsbYubiKeyDevice that stays live until unplugged.
      m.startUsbDiscovery(UsbConfiguration()) { device ->
        main.post { currentDevice = device }
        readSerialAndEmit(device, "usb")
        (device as? UsbYubiKeyDevice)?.setOnClosed {
          main.post {
            if (currentDevice === device) currentDevice = null
            emit("removed", "", "usb")
          }
        }
      }

      // NFC: best-effort. Needs a foreground Activity and an NFC radio; either
      // missing is fine — USB still works and isSupported() stays honest.
      try {
        val activity = (NitroModules.applicationContext as? ReactApplicationContext)?.currentActivity
        if (activity != null) {
          m.startNfcDiscovery(NfcConfiguration(), activity) { device ->
            main.post { currentDevice = device }
            readSerialAndEmit(device, "nfc")
            // An NFC tap is a transient session: when the tag leaves the field
            // we must emit `removed` so the JS layer relocks the PKM, exactly
            // like USB unplug. NfcYubiKeyDevice.remove(...) fires once the tag
            // is gone. Without this the 120s PKM window outlives the tap.
            (device as? NfcYubiKeyDevice)?.remove {
              main.post {
                if (currentDevice === device) currentDevice = null
                emit("removed", "", "nfc")
              }
            }
          }
        }
      } catch (_: NfcNotAvailable) {
        // no NFC on this device — ignore
      } catch (_: Throwable) {
        // any other NFC-start failure is non-fatal for USB-only use
      }
    }
  }

  override fun stopDiscovery() {
    main.post {
      if (!discovering) return@post
      val m = manager
      try { m?.stopUsbDiscovery() } catch (_: Throwable) {}
      try {
        val activity = (NitroModules.applicationContext as? ReactApplicationContext)?.currentActivity
        if (activity != null) m?.stopNfcDiscovery(activity)
      } catch (_: Throwable) {}
      currentDevice = null
      discovering = false
    }
  }

  override fun setKeyListener(listener: (String, String, String) -> Unit) {
    main.post { this.listener = listener }
  }

  override fun clearKeyListener() {
    main.post { this.listener = null }
  }

  private fun emit(eventType: String, serial: String, transport: String) {
    val l = listener ?: return
    main.post { l(eventType, serial, transport) }
  }

  /** Open a throwaway session just to read the serial for a connected event. */
  private fun readSerialAndEmit(device: YubiKeyDevice, transport: String) {
    device.requestConnection(SmartCardConnection::class.java) { result ->
      val serial = try {
        val piv = PivSession(result.value)
        piv.serialNumber.toString()
      } catch (_: Throwable) {
        ""
      }
      emit("connected", serial, transport)
    }
  }

  // ── operations ──

  override fun getKeyInfo(): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      val serial = piv.serialNumber
      val v = piv.version
      val retries = try { piv.pinAttempts } catch (_: Throwable) { -1 }
      "{\"serial\":\"$serial\",\"firmwareVersion\":\"${v.major}.${v.minor}.${v.micro}\",\"pinRetries\":$retries}"
    }
    return promise
  }

  override fun verifyPin(pin: String): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      try {
        piv.verifyPin(pin.toCharArray())
        "{\"ok\":true,\"retriesLeft\":null}"
      } catch (e: InvalidPinException) {
        // A wrong PIN is a normal, resolvable result for this probe (the spec
        // returns {ok, retriesLeft}); only transport faults reject.
        "{\"ok\":false,\"retriesLeft\":${e.attemptsRemaining}}"
      }
    }
    return promise
  }

  override fun changePin(oldPin: String, newPin: String): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      // PIV CHANGE REFERENCE DATA needs only the old PIN — NOT management-key
      // auth. Gating it on the management key would wrongly reject a key that
      // has a custom management key but a still-default PIN.
      try {
        piv.changePin(oldPin.toCharArray(), newPin.toCharArray())
        "{\"ok\":true}"
      } catch (e: InvalidPinException) {
        throw VaultException("pin-invalid", "retries=${e.attemptsRemaining}")
      }
    }
    return promise
  }

  override fun generateVaultKey(slot: Double, touchPolicy: String, pinPolicy: String): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      authenticateManagementKey(piv)
      // 3.2.0: generateKey(...) returns PublicKeyValues (generateKeyValues is
      // @Deprecated / removed).
      val pub = piv.generateKey(
        Slot.fromValue(slot.toInt()),
        KeyType.ECCP256,
        toPinPolicy(pinPolicy),
        toTouchPolicy(touchPolicy)
      )
      "{\"publicKey\":\"${(pub as PublicKeyValues.Ec).encodedPoint.toHex()}\"}"
    }
    return promise
  }

  override fun readVaultPublicKey(slot: Double): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      try {
        val meta = piv.getSlotMetadata(Slot.fromValue(slot.toInt()))
        val pub = meta.publicKeyValues as PublicKeyValues.Ec
        "{\"publicKey\":\"${pub.encodedPoint.toHex()}\"}"
      } catch (_: ApduException) {
        // Empty slot (reference-data-not-found) is not an error here.
        "{\"publicKey\":null}"
      }
    }
    return promise
  }

  override fun ecdh(slot: Double, pin: String, peerPublicKey: String): Promise<String> {
    val promise = Promise<String>()
    withPiv(promise) { piv ->
      piv.verifyPin(pin.toCharArray()) // pin-policy ONCE key gate; throws InvalidPinException on wrong PIN
      val peerBytes = hexToBytes(peerPublicKey)
      // 3.2.0: calculateSecret takes PublicKeyValues directly (the ECPublicKey
      // overload was removed) — do NOT convert to a java.security key.
      val peer = PublicKeyValues.Ec.fromEncodedPoint(EllipticCurveValues.SECP256R1, peerBytes)
      // TOUCH-gated when the key was generated with TouchPolicy.ALWAYS: this
      // blocks until the user taps the key, and surfaces as touch-timeout if
      // they never do (see mapError).
      val secret = piv.calculateSecret(Slot.fromValue(slot.toInt()), peer)
      "{\"secret\":\"${secret.toHex()}\"}"
    }
    return promise
  }

  override fun signEcdsa(slot: Double, pin: String, digest: String): Promise<String> {
    val promise = Promise<String>()
    // MUST be exactly 32 bytes: rawSignOrDecrypt silently TRUNCATES an over-long
    // EC payload (Arrays.copyOf to the key's 32-byte length) and left zero-pads a
    // short one, so an off-length digest signs the wrong message rather than
    // failing. Checked BEFORE any card command so a malformed digest never burns
    // a PIN retry.
    val digestBytes = try {
      hexToBytes(digest)
    } catch (t: Throwable) {
      promise.reject(vaultError("template-invalid", "digest must be hex"))
      return promise
    }
    if (digestBytes.size != 32) {
      promise.reject(vaultError("template-invalid", "digest must be exactly 32 bytes, got ${digestBytes.size}"))
      return promise
    }

    withPiv(promise) { piv ->
      // withPiv opens a FRESH PivSession per call, so the PIN must be verified
      // inside every operation — this is not redundant with an earlier verify.
      piv.verifyPin(pin.toCharArray())
      // TOUCH-gated by the slot's touch policy; a required-but-unmet touch
      // surfaces as SW 0x6982/0x6985, which mapError folds into touch-timeout.
      // rawSignOrDecrypt sends the digest verbatim (no local hashing/re-encoding)
      // and the card returns raw DER (SEQUENCE { r, s }), NOT low-S normalised —
      // returned here unmodified.
      val der = piv.rawSignOrDecrypt(Slot.fromValue(slot.toInt()), KeyType.ECCP256, digestBytes)
      "{\"signature\":\"${der.toHex()}\"}"
    }
    return promise
  }

  // ── helpers ──

  /**
   * Open a SmartCardConnection → PivSession against the held device, run
   * `block`, resolve; translate any failure to a VAULT_ERR rejection. No held
   * device means no key is on a reader → no-key.
   */
  private fun withPiv(promise: Promise<String>, block: (PivSession) -> String) {
    val device = currentDevice
    if (device == null) {
      promise.reject(vaultError("no-key", "no YubiKey present"))
      return
    }
    device.requestConnection(SmartCardConnection::class.java) { result ->
      try {
        val piv = PivSession(result.value) // result.value throws IOException if the connect failed
        promise.resolve(block(piv))
      } catch (t: Throwable) {
        promise.reject(mapError(t))
      }
    }
  }

  /**
   * Authenticate with the firmware-default management key so generateKey can
   * run. On yubikit-android 3.2.0 `authenticate(byte[])` reads the key's
   * algorithm from card metadata itself, so we pass only the 24-byte default
   * value (the same default for both the pre-5.7 TDES and fw >= 5.7 AES-192
   * cards). A rejection means the key has a custom management key we cannot
   * supply → mgmt-key-custom.
   */
  private fun authenticateManagementKey(piv: PivSession) {
    try {
      piv.authenticate(DEFAULT_MANAGEMENT_KEY)
    } catch (e: Throwable) {
      throw VaultException("mgmt-key-custom", "default management key rejected")
    }
  }

  private fun toTouchPolicy(p: String): TouchPolicy = when (p.lowercase()) {
    "always" -> TouchPolicy.ALWAYS
    "cached" -> TouchPolicy.CACHED
    "never" -> TouchPolicy.NEVER
    else -> TouchPolicy.DEFAULT
  }

  private fun toPinPolicy(p: String): PinPolicy = when (p.lowercase()) {
    "once" -> PinPolicy.ONCE
    "always" -> PinPolicy.ALWAYS
    "never" -> PinPolicy.NEVER
    else -> PinPolicy.DEFAULT
  }

  /** Small carrier so a code path can name its own VAULT_ERR code + detail. */
  private class VaultException(val code: String, val detail: String) : Exception("$code:$detail")

  private fun vaultError(code: String, detail: String): Throwable = Error("VAULT_ERR:$code:$detail")

  private fun mapError(t: Throwable): Throwable = when (t) {
    is VaultException -> Error("VAULT_ERR:${t.code}:${t.detail}")
    is InvalidPinException -> {
      val n = t.attemptsRemaining
      if (n <= 0) Error("VAULT_ERR:pin-locked:no attempts remaining")
      else Error("VAULT_ERR:pin-invalid:retries=$n")
    }
    is ApduException -> {
      when (t.sw.toInt() and 0xffff) {
        0x6983 -> Error("VAULT_ERR:pin-locked:authentication method blocked")
        0x6a88, 0x6a80 -> Error("VAULT_ERR:no-key:reference data not found")
        // 0x6982 (security status not satisfied) / 0x6985 (conditions not
        // satisfied) is what a required-but-unmet touch surfaces as over CCID.
        0x6982, 0x6985 -> Error("VAULT_ERR:touch-timeout:conditions of use not satisfied")
        else -> Error("VAULT_ERR:wrong-key:apdu 0x${Integer.toHexString(t.sw.toInt() and 0xffff)}")
      }
    }
    is IOException -> Error("VAULT_ERR:key-removed-mid-op:${t.message}")
    else -> Error("VAULT_ERR:wrong-key:${t.message}")
  }

  private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

  private fun hexToBytes(hex: String): ByteArray {
    val clean = hex.removePrefix("0x").removePrefix("0X")
    require(clean.length % 2 == 0) { "odd-length hex" }
    return ByteArray(clean.length / 2) {
      ((Character.digit(clean[it * 2], 16) shl 4) + Character.digit(clean[it * 2 + 1], 16)).toByte()
    }
  }

  companion object {
    /** Firmware-default PIV management key (0x0102…08 ×3, 24 bytes). */
    private val DEFAULT_MANAGEMENT_KEY = byteArrayOf(
      0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
      0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
      0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
    )
  }
}
